cylc/uiserver/resolvers.py

```diff
@@ -70,6 +80,107 @@ def check_cylc_version(version):
     return ret or out.strip() == version
 
 
+def _build_cmd(cmd: List, args: Dict) -> List:
+    """Add args to command.
+
+    Args:
+        cmd: A base command.
+        args: Args to append to base command.
+
+    Returns: An elaborated command.
+
+    Examples:
+        It adds one arg to a command:
+            >>> _build_cmd(['foo', 'bar'], {'set_baz': 'qux'})
+            ['foo', 'bar', '--set-baz', 'qux']
+
+        It adds one integer arg to a command:
+            >>> _build_cmd(['foo', 'bar'], {'set_baz': 42})
+            ['foo', 'bar', '--set-baz', '42']
+
+        It adds a list of the same arg to a command:
+            >>> _build_cmd(['foo', 'bar'], {'set_baz': ['qux', 'quiz']})
+            ['foo', 'bar', '--set-baz', 'qux', '--set-baz', 'quiz']
+
+        It doesn't append args == False:
+            >>> _build_cmd(['foo', 'bar'], {'set_baz': False})
+            ['foo', 'bar']
+    """
+    for key, value in args.items():
+        if value is False:
+            # don't add binary flags
+            continue
+        key = snake_to_kebab(key)
+        if not isinstance(value, list):
+            if isinstance(value, int):
+                # Any integer items need converting to strings:
+                value = str(value)
+            value = [value]
+        for item in value:
+            cmd.append(key)
+            if item is not True:
+                # don't provide values for binary flags
+                cmd.append(item)
+    return cmd
+
+
+def _schema_opts_to_api_opts(schema_opts: Dict) -> SimpleNamespace:
+    """Convert Schema opts to api Opts
+
+    Contains data SCHEMA_TO_API:
+        A mapping of schema options to functions in the form:
+        def func(schema_key, schema_value):
+            return (option_key, option _value)
+
+    Args:
+        schema_opts: Opts as described by the schema.
+
+    Returns:
+        Namespace for use as options.
+
+    TODO:
+        So far only `cylc clean` has been added. It may become necessary
+        to separate the SCHEMA_TO_API dict into GENERIC_SCHEMA_TO_API and
+        <COMMAND>_SCHEMA_TO_API, and add a kwarg indicating which
+        additional schema to use.
+
+    """
+    SCHEMA_TO_API: Dict[str, Union[Callable, None]] = {
+        'rm': lambda opt, value: ('rm_dirs', value),
+        'local_only': None,
+        'remote_only': None,
+        'debug':
+            lambda opt, value:
+                ('verbosity', 2) if value is True else ('verbosity', 0),
+        'no_timestamp': lambda opt, value: ('log_timestamp', not value),
+    }
```

Comment:
Should move this constant out of the function?

Reply:
I like it there because:
- I don't think it needs to be used elsewhere.
- Constant intimately linked with working of method - would you be asking me this if I'd made this function a class with a single method?
  ```python
  class SchemaOptsConverter:
      SCHEMA_TO_API = 'whateva'

      @staticmethod
      def _convert_schema_to_api_opts(schema_opts):
          ...
  ```

- I like the ability to document what it's for in the docstring.

I know that caps are normally reserved for module or class level variables, but I think it's appropriate here.

Reply:
It's not a constant, it gets redefined every time the function is run.

If you want to keep it here it should be lowercase, otherwise is should be moved outside the function.

Reply:
> would you be asking me this if I'd made this function a class with a single method?

No because that would be a constant.

Reply:
basically:

```python
MODULE_CONSTANT = 1

class C:
    CLASS_CONSTANT = 2

    def m(self):
        method_variable = 3

def f():
    function_variable = 4

```

Reply:
Basically Python code conventions (you will notice we don't do this elsewhere), it's hinted at in PEP8 but it's more general than that.